src/main/java/edu/byu/cs/autograder/compile/ModuleIndependenceVerifier.java

```diff
@@ -0,0 +1,59 @@
+package edu.byu.cs.autograder.compile;
+
+import edu.byu.cs.autograder.GradingContext;
+import edu.byu.cs.autograder.GradingException;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ModuleIndependenceVerifier implements StudentCodeInteractor {
+    @Override
+    public void interact(GradingContext context, StudentCodeReader reader) throws GradingException {
+        Set<File> serverFiles = reader.filesMatching(".*server/src/main/java/.*").collect(Collectors.toSet());
+        Path serverPath = context.stageRepo().toPath().resolve("server/src/main/java");
+        Set<String> serverClasses = qualifiedNames(serverFiles, serverPath);
+
+        Set<File> clientFiles = reader.filesMatching(".*client/src/main/java/.*").collect(Collectors.toSet());
+        Path clientPath = context.stageRepo().toPath().resolve("client/src/main/java");
+        Set<String> clientClasses = qualifiedNames(clientFiles, clientPath);
+
+        try {
+            checkImports(context, reader, serverFiles, clientClasses);
+            checkImports(context, reader, clientFiles, serverClasses);
+        } catch (IOException e) {
+            throw new GradingException("Unable to verify module independence", e);
+        }
+    }
+
+    private Set<String> qualifiedNames(Set<File> files, Path relativeTo) {
+        Set<String> classes = new HashSet<>();
+        for(File file : files) {
+            String fileName = relativeTo.relativize(file.toPath()).toString();
+            if(!fileName.contains(".java")) continue;
+            classes.add(fileName.substring(0, fileName.indexOf('.')).replace('/', '.'));
+        }
+        return classes;
+    }
+
+    private void checkImports(GradingContext context, StudentCodeReader reader, Set<File> files, Set<String> classes)
+            throws IOException {
+        for(File file : files) {
+            List<String> contents = reader.getFileContents(file);
+            for (String line : contents) {
+                if (line.matches("(?:abstract )?class|interface|record")) break;
+                for (String classImport : classes) {
```

Comment:
Currently, you're iterating over all classes to see if any of them match the format. However, the format is fixed. It's always `import <CLASS_NAME>;` It would be a lot more effective to check for set inclusion by using a HashSet of all the known classes.

That makes this an N*M operation where N is the number of import lines and M is the number of classes. You could make it an N + M operation instead which would be basically processing each line & import once.

You could do either of the following to improve the efficiency:
1. Prepare a `HashSet` in advance with the line contents pre-appended, or
2. Extract the class name from the import statement and then compare it against the `HashSet`

Reply:
Good points. Firstly, this would catch static imports. Do you think that's a big deal?

Reply:
How does that look? I think I found a way to do both

Reply:
> `if (packages.contains(packageImport))`

I think it's perfect 👌